### PR TITLE
Add possibility to use custom `User Data`

### PR DESCRIPTION
**Pull request by SweetOps**
## What

* Add possibility to use custom `User Data`
* Update README
* Add template for PR

## Why

* Some `AMIs` doesn't support classic `User Data`

### CODE REVIEW

**Comment by SweetOps** [comment]
has been tested with following manifest 
```terraform
data "aws_ami" "pfsense" {
  most_recent = "true"

  filter {
    name   = "description"
    values = ["Netgate pfSense Certified 2.3.4"]
  }
}

data "template_file" "init" {
  template = "${file("1.sh")}"
}

module "admin_tier" {
  source                        = "cloudposse/terraform-aws-ec2-instance"
  ssh_key_pair                  = "sweetops"
  ami                           = "${data.aws_ami.pfsense.image_id}"
  vpc_id                        = "vpc-someid"
  subnet                        = "subnet-someid"
  associate_public_ip_address   = "true"
  create_default_security_group = "false"
  name                          = "pf"
  namespace                     = "cp"
  stage                         = "dev"
  user_data                     = ["${data.template_file.init.rendered}"]
}

module "admin_tier1" {
  source                        = "cloudposse/terraform-aws-ec2-instance"
  ssh_key_pair                  = "sweetops"
  ami                           = "${data.aws_ami.pfsense.image_id}"
  vpc_id                        = "vpc-someid"
  subnet                        = "subnet-someid"
  associate_public_ip_address   = "true"
  create_default_security_group = "false"
  name                          = "pf1"
  namespace                     = "cp"
  stage                         = "dev"
  custom_user_data              = ["${data.template_file.init.rendered}"]
}

```

File: main.tf [hunk 1]
```diff
@@ -7,6 +7,7 @@ locals {
   availability_zone    = "${var.availability_zone != "" ? var.availability_zone : data.aws_subnet.default.availability_zone}"
   ami                  = "${var.ami != "" ? var.ami : data.aws_ami.default.image_id}"
   root_volume_type     = "${var.root_volume_type != "" ? var.root_volume_type : data.aws_ami.info.root_device_type}"
+  user_data            = "${length(var.custom_user_data) > 0 ? join("", var.custom_user_data) : join("", data.template_file.user_data.*.rendered)}"
```
Comment (comeanother) [inline main.tf]: Should we use `var.custom_user_data` instead of `join("", var.custom_user_data) `?

File: user_data.tf [hunk 2]
```diff
@@ -1,4 +1,5 @@
 data "template_file" "user_data" {
+  count    = "${local.instance_count && length(var.custom_user_data) == 0 ? 1 : 0}"
```
Comment (comeanother) [inline user_data.tf]: We should use `local.instance_count == 0`.

File: outputs.tf [hunk 3]
```diff
@@ -43,11 +43,16 @@ output "alarm" {
...
 }
 
-output "additional_eni_ids" {
-  description = "Map of ENI with EIP"
+output "additional_eni_public" {
+  description = "Map IDs of ENI with public IPs"
   value       = "${zipmap(aws_network_interface.additional.*.id, aws_eip.additional.*.public_ip)}"
 }
 
+output "additional_eni_private" {
```
Comment (comeanother) [inline outputs.tf]: Why do we need this for?

File: main.tf (hunk 1 above)
Comment (SweetOps) [reply main.tf]: in this case will be error during plan\apply, because trueval\falseeval should be string

File: outputs.tf (hunk 3 above)
Comment (SweetOps) [reply outputs.tf]: for example it need in sth for template rendering

**Comment by SweetOps** [comment]
screenshot 2017-11-27 20 40 25

**Comment by osterman** [comment]
We are overloading this module. Enforcing a convention on the user data that including using `github_authorized_keys` is too restrictive. It's easy enough to pass these when invoking the module, especially since we have a module to generate user data for ghk. 

Closing this in favor of https://github.com/cloudposse/terraform-aws-ec2-instance/pull/22

**Pull request closed without merging**